//
//protoc --cpp_out . FromServer.proto
//C:\code\libraries\vcpkg\installed\x64-windows\tools\protobuf\protoc.exe --cpp_out . FromServer.proto
//ln -s /home/duffyj/code/jde/ApplicationServer/source/types/proto/FromServer.proto ../proto/appFromServer.proto
//npx pbjs -r app_from_server -t static-module -w es6 -o src/app/proto/appFromServer.js ../proto/appFromServer.proto & npx pbts -o src/app/proto/appFromServer.d.ts src/app/proto/appFromServer.js
syntax = "proto3";
package Jde.ApplicationServer.Web.FromServer;
//import "google/protobuf/timestamp.proto";

message Acknowledgement
{
	uint32 Id=1;
}

enum ELogLevel
{
	Trace = 0;
	Debug = 1;
	Information = 2;
	Warning = 3;
	Error = 4;
	Critical = 5;
	None = 6;
}

message Application
{
	uint32 Id=1;
	string Name=2;
	ELogLevel DbLevel=3;
	ELogLevel FileLevel=4;
}

message Applications
{
	repeated Application Values=1;
}

message Status
{
	uint32 ApplicationId=1;
	uint32 InstanceId=2;
	string HostName=3;
	uint32 StartTime=4;
	uint64 Memory=5;
	uint32 LogCount=6;
	double Cpu=7;
	ELogLevel DBLogLevel=8;
	ELogLevel FileLogLevel=9;
	repeated string Values=10;
}

message Statuses
{
	repeated Status Values=1;
}

message TraceMessage
{
	uint64 InstanceId=1;
	uint64 Time=2;
	ELogLevel Level=3;
	uint32 MessageId=4;
	uint32 FileId=5;
	uint32 FunctionId=6;
	uint32 LineNumber=7;
	uint32 UserId=8;
	uint64 ThreadId=9;
	repeated string Variables=10;
}

message Traces
{
	uint32 ApplicationId=1;
	repeated TraceMessage Values=2;
}

message ApplicationString
{
	uint32 StringRequestType=1;
	uint32 Id=2;
	string Value=3;
}

message ApplicationStrings
{
	uint32 RequestId=1;
	uint32 ApplicationId=2;
	repeated ApplicationString Values=3;
}

message ErrorMessage
{
	uint32 RequestId=1;
	string Message=2;
}

message Custom
{
	uint32 RequestId=1;
	bytes Message=2;
}

message Complete
{
	uint32 RequestId=1;
}

message MessageUnion
{
	oneof Value
	{
		Traces Traces=1;
		Statuses Statuses=2;
		Acknowledgement Acknowledgement=3;
		ApplicationStrings Strings=4;
		Applications Applications=5;
		ErrorMessage Error=6;
		Custom Custom=7;
		Complete Complete=8;
	}
}

message Transmission
{
	repeated MessageUnion Messages=1;
}
